lib/spack/spack/solver/asp.py

```diff
@@ -2076,6 +2076,10 @@ def setup(self, driver, specs, reuse=None):
 
         # get list of all possible dependencies
         self.possible_virtuals = set(x.name for x in specs if x.virtual)
+        for spec in specs:
+            if not spec.virtual:
+                for provided in spec.package_class.provided:
+                    self.possible_virtuals.add(provided.name)
```

Comment:
I'll give a thorough review tomorrow, but curious why this was needed.

Reply:
When I introduced a choice rule for impose, it gave the concretizer the power to impose selective attributes about nodes without them being part of the dag, which was nonsensical, so I introduced 

```
error(1, "Unexpected attribute {0}/{1}", Package, X) :- attr(X, Package), not attr("virtual_node", Package), not external(Package), not attr("node", Package).
```

i.e. things must be part of the dag for anny attribute to be set on them. Interestingly, this led to the discovery that in some circumstances there are "entities" in our DAG which are not nodes or virtual nodes but still have attributes set on them. I think specifically it is providers that are specified as root specs (but you can know for sure by removing this, which will cause test faulures), so logic like this was added to ensure that everything that we set attrs on *is* in fact a node or virtual node (or external).